bin/weewx/drivers/ws1.py

```diff
@@ -221,6 +203,209 @@ def _decode(s, multiplier=None, neg=False):
+                    buf = buf[buf.find('!!'):]
+                    if DEBUG_READ: loginf("(found!) buf: %s" % buf)
+                    break
+            # Add the rest of the record
+            try:
+                buf += self.net_socket.recv(
+                    PACKET_SIZE - len(buf), socket.MSG_WAITALL)
+            except (socket.error, socket.timeout), ex:
+                raise weewx.WeeWxIOError(ex)
+        else:
+            # Keep receiving data until we find an exclamation point or two
+            try:
+                buf = self.net_socket.recv(2, socket.MSG_WAITALL)
+            except (socket.error, socket.timeout), ex:
+                raise weewx.WeeWxIOError(ex)
+            while True:
+                if buf == '\r\n':
+                    # CRLF is expected
+                    if DEBUG_READ: loginf("buf is CRLF")
+                    buf = ''
+                    break
+                elif '!' in buf:
+                    excmks = buf.count('!')
+                    # Assuming exclamation points are at the end of the buffer
+                    buf = buf[len(buf) - excmks:]
+                    if DEBUG_READ: loginf("buf has %d exclamation points." % (
+                        excmks))
+                    break
+                else:
+                    try:
+                        buf = self.net_socket.recv(2, socket.MSG_WAITALL)
+                    except (socket.error, socket.timeout), ex:
+                        raise weewx.WeeWxIOError(ex)
+                    if DEBUG_READ: loginf("buf: %s" % ' '.join(
+                        ['%02X' % ord(bc) for bc in buf]))
+            try:
+                buf += self.net_socket.recv(
+                    PACKET_SIZE - len(buf), socket.MSG_WAITALL)
+            except (socket.error, socket.timeout), ex:
+                raise weewx.WeeWxIOError(ex)
+        if DEBUG_READ: loginf("buf: %s" % buf)
+        # This code assumes CRLF will be transmitted at the end of each record,
+        # which may not always be the case. See Matthew Wall's comment on
+        # GitHub here:
+        # https://github.com/weewx/weewx/pull/86#issuecomment-166716509
+
+        # try:
+        #     self.net_socket.recv(2, socket.MSG_WAITALL)  # CRLF
+        # except (socket.error, socket.timeout), ex:
+        #     raise weewx.WeeWxIOError(ex)
+        buf.strip()
+        return buf
+
+    def get_readings_with_retry(self, max_tries=5, retry_wait=10):
+        for ntries in range(0, max_tries):
+            try:
+                buf = self.get_readings()
+                StationData.validate_string(buf)
+                return buf
+            except (weewx.WeeWxIOError), e:
+                loginf("Failed to get data for some reason: %s" % e)
+                self.rec_start = False
+
+                # WeeWx IO Errors may not always occur because of invalid data
+                # DEBUG_READ = True
+
+                # This causes an exception
```

Comment:
declare buf outside the try block, e.g., buf = ''.  be sure to wrap the socket errors into WeeWxIO errors otherwise weewx will die.  there should be a generic socket.error or some such which you can catch then raise WeeWXIOError(e)